nipc-openapi/NIPC.yaml

```diff
@@ -0,0 +1,1619 @@
+      requestBody:
+        content:
+          application/json:
+            schema:
+              $ref: './protocolmaps/ProtocolMap.yaml#/components/schemas/ProtocolMap-Property'
+        required: true
+      responses:
+        '200':
+          description: Success
+          content:
+            application/json:
+              schema:
+                $ref: '#/components/schemas/SuccessResponse'          
+        '400':
+          description: Bad request
+        '401':
+          description: Unauthorized
+        '405':
+          description: Invalid request 
+        '500':
+          description: Server-side failure
+          content:
+            application/json:
+              schema:
+                allOf:
+                  - $ref: '#/components/schemas/FailureResponse' 
+                  - $ref: '#/components/schemas/EventErrorCodes'
+ 
+### Registrations
+  /registration/sdfobject/{sdfobject}:
+    post:
+      tags:
+        - NIPC registration APIs
+      summary: Register an sdfObject
+      description: |-
+        Register an sdfObject, including Properties, Events and actions
+      operationId: registerSdfObject
+      parameters:
+        - name: sdfobject
+          in: path
+          description: sdfObject name
+          required: true
+          schema:
+            type: string
+            example: "temperature-sensor"
+      requestBody:
+        content:
+          application/json:
+            schema:
+              allOf:
+                - $ref: '#/components/schemas/SdfObject'
+                - $ref: '#/components/schemas/Id'
+        required: true
+      responses:
+        '200':
+          description: Success
+          content:
+            application/json:
+              schema:
+                allOf:
+                  - $ref: '#/components/schemas/SuccessResponse'
+                  - $ref: '#/components/schemas/SdfObject'
+                  - $ref: '#/components/schemas/Id'
+        '400':
+          description: Bad request
+        '401':
+          description: Unauthorized
+        '405':
+          description: Invalid request
+        '500':
+          description: Server-side failure
+          content:
+            application/json:
+              schema:
+                allOf:
+                  - $ref: '#/components/schemas/FailureResponse' 
+                  - $ref: '#/components/schemas/RegistrationErrorCodes'
+ 
+    put:
+      tags:
+        - NIPC registration APIs
+      summary: Update an sdfObject
+      description: |-
+        Update an sdfObject, including Properties, Events and actions
+      operationId: updateSdfObject
+      parameters:
+        - name: sdfobject
+          in: path
+          description: sdfObject name
+          required: true
+          schema:
+            type: string
+            example: "temperature-sensor"
+      requestBody:
+        content:
+          application/json:
+            schema:
+              allOf:
```

Comment:
OpenAPI preview shows example request here as
```json
{
  "namespace": {
    "heartrate": "https://example.com/heartrate"
  },
  "defaultNamespace": "heartrate",
  "sdfObject": {
    "additionalProp1": {
      "sdfProperty": {
        "additionalProp1": {
          "protocolMap": {
            "ble": {
              "serviceID": "12345678-1234-5678-1234-56789abcdef4",
              "characteristicID": "12345678-1234-5678-1234-56789abcdef4"
            }
          }
        },
        "additionalProp2": {
          "protocolMap": {
            "ble": {
              "serviceID": "12345678-1234-5678-1234-56789abcdef4",
              "characteristicID": "12345678-1234-5678-1234-56789abcdef4"
            }
          }
        },
        "additionalProp3": {
          "protocolMap": {
            "ble": {
              "serviceID": "12345678-1234-5678-1234-56789abcdef4",
              "characteristicID": "12345678-1234-5678-1234-56789abcdef4"
            }
          }
        }
      }
    },
    "additionalProp2": {
      "sdfProperty": {
        "additionalProp1": {
          "protocolMap": {
            "ble": {
              "serviceID": "12345678-1234-5678-1234-56789abcdef4",
              "characteristicID": "12345678-1234-5678-1234-56789abcdef4"
            }
          }
        },
        "additionalProp2": {
          "protocolMap": {
            "ble": {
              "serviceID": "12345678-1234-5678-1234-56789abcdef4",
              "characteristicID": "12345678-1234-5678-1234-56789abcdef4"
            }
          }
        },
        "additionalProp3": {
          "protocolMap": {
            "ble": {
              "serviceID": "12345678-1234-5678-1234-56789abcdef4",
              "characteristicID": "12345678-1234-5678-1234-56789abcdef4"
            }
          }
        }
      }
    },
    "additionalProp3": {
      "sdfProperty": {
        "additionalProp1": {
          "protocolMap": {
            "ble": {
              "serviceID": "12345678-1234-5678-1234-56789abcdef4",
              "characteristicID": "12345678-1234-5678-1234-56789abcdef4"
            }
          }
        },
        "additionalProp2": {
          "protocolMap": {
            "ble": {
              "serviceID": "12345678-1234-5678-1234-56789abcdef4",
              "characteristicID": "12345678-1234-5678-1234-56789abcdef4"
            }
          }
        },
        "additionalProp3": {
          "protocolMap": {
            "ble": {
              "serviceID": "12345678-1234-5678-1234-56789abcdef4",
              "characteristicID": "12345678-1234-5678-1234-56789abcdef4"
            }
          }
        }
      }
    }
  },
  "id": "12345678-1234-5678-1234-56789abcdef4"
}
```

The `id` field is not valid SDF, but this could be a reasonable pattern for instantiation of SDF models (to be discussed).

The "additionalPropN" are somewhat misleading names for objects. And probably the intended design was to have just one object and three properties in that instead of three objects?

Reply:
The preview generates this example automatically and uses `additionalPropN` to signify any property where the name can be any string. Let me update the yaml with a proper example.

Reply:
I have updated the example so it shouldn't say "additionalPropN" anymore.

Reply:
The id field is required to map the sdfobject ot an actual device id. 
An alternative would be to move {id} to the path

Reply:
> I have updated the example so it shouldn't say "additionalPropN" anymore.

Thanks! Now the example looks better

Reply:
> The id field is required to map the sdfobject ot an actual device id. An alternative would be to move {id} to the path

I'd assume the same SDF model can be shared by multiple devices of same type. Would it make sense to have separate model registration API endpoint and then a registration endpoint for the devices that can then refer to the (already registered) models? Or you could perhaps combine both by having the id in the path so that the model remains re-usable.

Reply:
@akeranen id is either a device id or group id, so the sdf model could be registered for a device or a group of devices

Reply:
> @akeranen id is either a device id or group id, so the sdf model could be registered for a device or a group of devices

OK; but I guess that's orthogonal issue wrt re-using the models? So you could get re-use with group-ids or by devices with different IDs referring to the same model?

Reply:
Yes, something like that. I haven't thought through the design and flow but this kind of capability seems useful. Another way to get re-use is to refer to existing (already registered) models with `sdfRef` (but that creates new models which may or may not be intended).